### ISSUE TITLE
Add Windows Authentication support to WiltonDB

### ISSUE THREAD

**Issue by staticlibs** [wiltondb/wiltondb#18]
Babelfish supports [GSSAPI](https://www.postgresql.org/docs/15/gssapi-auth.html), but it is not included with WiltonDB. In theory GSSAPI (MIT Kerberos) may be used with Windows Kerberos (Active Directory) when running on Windows as the Kerberos protocol is supposed to be compatible between MIT Kerberos and Windows Kerberos. But in practice PostgreSQL on Windows does not use GSSAPI to talk with Active Directory - uses [SSPI](https://www.postgresql.org/docs/15/sspi-auth.html) (Windows Kerberos with NTLM fallback) instead, and SSPI  part is not implemented in Babelfish for TDS protocol.

This issue is to explore the possibilities to implement SSPI for TDS connections in WiltonDB.

**Comment by psrinivasa2** [comment]
@staticlibs can you please let us know where we on this?

**Comment by staticlibs** [comment]
@psrinivasa2 , there has been no work done yet on this issue.

To clarify the situation: I am currently working on [#2137](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/2137) and need to complete it before taking on another major tasks (like SSPI support for Windows Auth). I do think that there are more potential users of Babelfish on Windows who are interested in Windows Auth (because many MSSQL users rely on it), than in legacy jTDS driver support. Just my work on jTDS is (partially) sponsored by the company which needs it - so this task takes the preference.

After [#2137](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/2137) is completed (I think it is close to completion) I intend to look into 2 minor related TDS issues ([#2402](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/2402) and #14 - not sure if there will be any changes for them) and prepare updated WiltonDB packages with jTDS support. After that I will be able to look into SSPI.

**Comment by psrinivasa2** [comment]
Hi @staticlibs,
I know you may be busy with concurrent works, but currently this is a blocker for us in integrating wiltonDB with our application, can you please let us know is this started or yet to be?

**Comment by staticlibs** [comment]
@psrinivasa2, yes, the work on this has started now, I am currently researching how SSPI works in Postgres and how GSSAPI support is implemented in Babelfish on Linux. I will update this issue with details once the steps to implement are more clear. One point that is completely unclear now is how to deal with SQL Browser service, SSPI connections use it to lookup the instance port. It seems that some kind of replacement for SQL Browser needs to be implemented for WiltonDB.

**Comment by psrinivasa2** [comment]
thanks @staticlibs

**Comment by staticlibs** [comment]
As an update on this, I've got a working prototype of Windows Auth on TDS connections. It is tested to work with ADO.NET (`Integrated Security=True`), mssql-jdbc, SSMS and bcp. It currently does not work with wdb_transfer (wdb_transfer itself needs to be updated).

@psrinivasa2 , some details and a few questions to you on the projected usage:

1. SQL Browser appeared to be not required for Windows Auth, all MSFT client libs/apps support specifying TCP port directly instead of instance name.

2. current server-side implementation in WiltonDB and all the MSFT client libs/apps listed above use  [Negotiate](https://learn.microsoft.com/en-us/windows/win32/secauthn/microsoft-negotiate) mode for auth. It is supposed to work transparently with both NTLM and Kerberos protocols, but was only tested with NTLM (that does not require AD setup). Would it be possible to use NTLM in your case? I know it may be prohibited by IT as "less secure" in some environments. I won't make any promises about Kerberos support - AD setup for its testing is unreasonably complex.

3. what are your requirements to the DB logins that are selected for incoming OS users? Can all incoming users be mapped to a single DB login, or do you require support for [username maps](https://www.postgresql.org/docs/15/auth-username-maps.html)? I would like to keep this part as simple as possible for initial versions. And also would like to avoid `pg_ident.conf` file usage (used by SSPI/GSS for Postgres connections), as it is easy to mess up. If it is acceptable to have a single regex for incoming OS usernames mapped to a single DB login - it will be easier to implement.

**Comment by psrinivasa2** [comment]
I will do check on it and come back @staticlibs

**Comment by psrinivasa2** [comment]
@staticlibs ,

1) For the First point, we are already using only server name in our config file, since during deployment in different machine's we no need to know about port number if it is  different for different customer machine's. Is it possible to have the same behavior?
![image](https://github.com/wiltondb/wiltondb/assets/89133439/64a2101e-e797-4b26-bb75-da9bd664ef4f)

2) For the second one, when I check the authentication scheme in SQL server which we are using, it says NTLM. So we already using NTLM, it can be same for the time being
![image](https://github.com/wiltondb/wiltondb/assets/89133439/da07757a-427b-439f-bfbc-44410cbaea7a)

3) For the third one, Based on my understanding in SQL, for every windows OS user windows authentication login , the SQL automatically create a equivalent DB login. In such a way each OS user will have different DB login's. This is will be helpful when it comes to auditing. how difficult it is to mimic the same SQL behavior here? Please correct me if my understanding is wrong.

**Comment by psrinivasa2** [comment]
@staticlibs for the first one, we can go-ahead with specifying TCP port with server name, since even in SQL instance name will work for default port alone, if it is changed, we need to mention the port number with instance name, I believe.

**Comment by staticlibs** [comment]
@psrinivasa2 , I will follow up on point 3 and will prepare a release candidate MSI build for you to try in a day or two.

**Comment by psrinivasa2** [comment]
@staticlibs  Regarding the point, 'It currently does not work with wdb_transfer (wdb_transfer itself needs to be updated).'

Sometimes back you mentioned that you have implemented windows auth for wdb_trabsfer which I have not tested yet since I faced during that time and left there, do you mean to say it got broken because of recent change?

**Comment by staticlibs** [comment]
@psrinivasa2 , I will follow up on wdb_transfer tomorrow.

**Comment by psrinivasa2** [comment]
thanks @staticlibs

**Comment by staticlibs** [comment]
@psrinivasa2 

> @staticlibs Regarding the point, 'It currently does not work with wdb_transfer (wdb_transfer itself needs to be updated).'
> 
> Sometimes back you mentioned that you have implemented windows auth for wdb_trabsfer which I have not tested yet since I faced a error during that time and left there, do you mean to say it got broken because of recent change?
> 
> [babelfish-for-postgresql/babelfish-for-postgresql#15 (reply in thread)](https://github.com/babelfish-for-postgresql/babelfish-for-postgresql/discussions/15#discussioncomment-9742631)

About the Windows Auth support in `wdb_transfer` - it was added in `1.2.0`. There is no custom auth code there, Windows Auth is supported by (fairly popular) [underlying lib](https://github.com/prisma/tiberius). What is specific about this lib, is that it relies on SQL Browser for Windows Auth. It takes instance name instead of port number (this is reflected in `wdb_transfer` UI) and looks up the port number. This bit was failing for you [in this comment](https://github.com/babelfish-for-postgresql/babelfish-for-postgresql/discussions/15#discussioncomment-9747381), I've [replied to it](https://github.com/babelfish-for-postgresql/babelfish-for-postgresql/discussions/15#discussioncomment-9747993) providing python script to troubleshoot. There was no activity on this after it, so I've closed #17 . If you still have problems with Windows Auth in `wdb_transfer` - please just reopen #17 and add details there.

All the above is about using `wdb_transfer` with MSSQL. With upcoming Windows Auth in WiltonDB - there is no support for SQL Browser - it seems to be unnecessary. All MSFT client apps/DB drivers support specifying `machine_name,port_number` instead of `machine_name\instance_name`. And port number is `1433` by default, so can be omitted in that case. This support for specifying the ports number for Windows Auth is currently lacking in `wdb_transfer` UI and underlying lib (invoked `bcp` supports it fine) - this needs to be improved in `wdb_transfer` after WiltonDB update is out.

**Comment by psrinivasa2** [comment]
okay got it Alex @staticlibs , I will need to check two issue with wdb_transfer from my end 1) windows authentication 2) duplicate key value violates unique constraint "babelfish_view_def_pkey". Once I complete my analysis area, I need to deep dive into these two and check it.

**Comment by staticlibs** [comment]
@psrinivasa2 , I've prepared a [release candidate](https://github.com/wiltondb/wiltondb/releases/tag/3.3-10-15-0-rc1) MSI with Windows Auth support included, see details about its usage on draft [Windows Authentication wiki page](https://github.com/wiltondb/wiltondb/wiki/Windows-Authentication-%5BDRAFT%5D#setting-up-windows-authentication-in-wiltondb).

**Comment by psrinivasa2** [comment]
Thanks much Alex, @staticlibs. I will check on it.

**Comment by psrinivasa2** [comment]
Hi Alex, @staticlibs,
This installer seems to be common DB login rather than SQL behavior of different automatically created DB logins for different windows users. Any update on that point?

![image](https://github.com/user-attachments/assets/c05d4268-0544-458a-9ac4-2bf76f46a278)

![image](https://github.com/user-attachments/assets/5125a227-9b85-4936-a5d1-461b79260e9d)

![image](https://github.com/user-attachments/assets/c21001bd-352d-454e-a594-f4669780bf61)

**Comment by psrinivasa2** [comment]
Thanks Alex @staticlibs , Can you bit elaborate on the point 'the intention is to use wilton_winauth login for all SSPI logins and to provide the ability to fetch OS user name for audit purposes' how can this be done?

Moreover, when I login as windows auth in SSMS (DESKTOP-JS80C9O, 2433) based on the instruction given in the document, It is successful. Further when I try to connect to a database via above connection (DESKTOP-JS80C9O, 2433) which was created in localhost,2433, I am getting below error, any idea on it Alex?


![image](https://github.com/user-attachments/assets/757a3f73-040b-48f0-b3cd-143ed01b5440)

**Comment by staticlibs** [comment]
@psrinivasa2 , the following query can be used from SSPI session to get a name of OS user (also described in [this section on a draft page](https://github.com/wiltondb/wiltondb/wiki/Windows-Authentication-%5BDRAFT%5D#user-sessions-audit)):

```tsql
select setting from pg_settings where name = 'wilton_winauth_os_user'
```
About connecting to T-SQL databases in SSPI session - you need to give `wilton_winauth` login appropriate permissions. For example, make this login a sysadmin with the following:

```tsql
alter server role sysadmin add member wilton_winauth
```

**Comment by psrinivasa2** [comment]
Thanks, Alex @staticlibs

**Comment by staticlibs** [comment]
@psrinivasa2 , yes, going to update `wdb_transfer` (to support  Windows Auth without instance name) after debug MSI for #24 is out.

**Comment by psrinivasa2** [comment]
Hi Alex @staticlibs  Where are we on this for above wdb_transfer windows authentication change

**Comment by staticlibs** [comment]
@psrinivasa2 , `wdb_transfer` UI changes are in progress, update should be ready in a few days.

**Comment by psrinivasa2** [comment]
Noted, thanks Alex

**Comment by psrinivasa2** [comment]
Hi Alex @staticlibs, 
Good Day!

Can you please share the update on above?

**Comment by staticlibs** [comment]
@psrinivasa2 , `wdb_transfer` changes are in progress, there are complications with underlying DB access lib that relies on SQL Browser, should be solvable, need few more days for this.

**Comment by psrinivasa2** [comment]
Okay Alex @staticlibs

**Comment by staticlibs** [comment]
As an update to this, I've got to the bottom of the problem with [underlying DB lib](https://github.com/prisma/tiberius). There is a bug in Windows auth impl in it, [submitted a fix](https://github.com/prisma/tiberius/pull/351) there. With modified Tiberius lib Windows auth is working with WiltonDB. So only UI changes are remaining now for `wdb_transfer`.

**Comment by psrinivasa2** [comment]
Thanks Alex, hope you will consider about the below points in transfer tool as well which you have shared separately with me

I can add the following logic to SSPI login implementation:
1. if DB login exists with the same name as OS user - this DB login is used
2. otherwise wilton_winauth DB login is used
3. if neither of DB logins exist - login process is failed

**Comment by psrinivasa2** [comment]
A point triggered on my mind, Till now we tried windows auth via application connection string and SSMS. Is there away to connect to pgAdmin tool as windows user?

**Comment by staticlibs** [comment]
@psrinivasa2 , yes, I can confirm that Windows auth on Postgres connection works fine with WiltonDB. There are no Babelfish specifics about it, normal Postgres SSPI setup can be used. Basically you need to set `auth-method` to `sspi` in `pg_hba.conf`, create a user map in `pg_ident.conf` and use "Kerberos Authentication" option in PgAdmin. Docs [reference](https://www.postgresql.org/docs/15/client-authentication.html).

Though having Windows auth on the same WiltonDB instance on both TDS and Postgres connections may requires non-trivial setup in `pg_hba.conf`. I suggest to not mix both Postgres and TDS Windows auth on the same instance. TDS Windows auth does not require any changes to `pg_hba.conf`.

**Comment by staticlibs** [comment]
@psrinivasa2 , I've added support for 1-to-1 OS user to DB login mapping in [11-16-0-rc2](https://github.com/wiltondb/wiltondb/releases/tag/3.3-11-16-0-rc2).

The implementation uses the fact, that `create login [domain\username] from windows` is supported in Babelfish for GSS TDS login. Such `from windows` logins are stored in Postgres (`pg_roles`, `pg_authid`) and Babelfish (`babelfish_authid_login_ext`) catalogs in `username@DOMAIN` Kerberos format. Due to this storage and processing details, such login names are not case-sensitive, support only ASCII characters and cannot include `@` symbol.

Besides  the `@` (that must be replaced with `_` in all T-SQL commands), there are no other transformation requirements. I've updated corresponding section [in wiki](https://github.com/wiltondb/wiltondb/wiki/Windows-Authentication#setting-up-windows-authentication-in-wiltondb).

**Comment by staticlibs** [comment]
@psrinivasa2 , I've updated `wdb_transfer` to support Windows auth with WiltonDB - [1.3.0 release](https://github.com/wiltondb/wdb_transfer/releases/tag/1.3.0). It will require SQLBrowser service with MSSQL only when "Connect to named MSSQL instance" checkbox is enabled.

**Comment by psrinivasa2** [comment]
Hi Alex @staticlibs, 
I tested both WiltonDB and transfer tool in high level, I can you confirm that It works fine. I will keep you posted, If I find any issue in it.

**Comment by psrinivasa2** [comment]
![image](https://github.com/user-attachments/assets/e7426565-b62f-4946-8b57-04120e3e6761)

Any idea on this Alex?
_only sysadmin role is supported in ALTER SERVER ROLE statement_

**Comment by staticlibs** [comment]
@psrinivasa2 , yes, only "sysadmin" role is implemented in Babelfish. Other roles like "securityadmin" are not supported. It is documented in [this wiki section](https://github.com/wiltondb/wiltondb/wiki/Logins-and-users#creating-login-with-sysadmin-privileges).

**Comment by psrinivasa2** [comment]
Hi Alex,
The moment we create 'create login [domain\username] from windows', what is the role attached with it?

![image](https://github.com/user-attachments/assets/22b712e8-22d4-463d-af25-894d8152c01f)
Apart server level permission, what are the database level permissions are possible?

**Comment by staticlibs** [comment]
@psrinivasa2 , by default new login has a "guest" role and can create objects in "master.guest" schema.

In a particular database users created for logins can be granted permissions like "select" or "insert" for particular tables or views ([wiki link](https://github.com/wiltondb/wiltondb/wiki/Logins-and-users#granting-user-permissions-for-particular-tables)).

**Comment by psrinivasa2** [comment]
Thanks Alex, @staticlibs.

My understanding is when I execute ''_create login [domain\john] from windows_',  it will be guest role access to only 'master.guest' schema. I need to create the database user for this login and assign other further permission's for relevant database like wilton database as whole or particular schema or table, etc

When I execute ''_create login [domain\john] from windows_' and '_alter server role serveradmin add member  [domain\john]_', no need to create separate database user as this by default has all permission's.

is this correct?

**Comment by staticlibs** [comment]
@psrinivasa2 , yes, this is correct.

Overall there are 3 main options for permissions management: 

 - sysadmin login
 - T-SQL DB owner login
 - user inside the DB with fine grained permissions

**Comment by psrinivasa2** [comment]
can you bit elaborate on 'T-SQL DB owner login'?

**Comment by staticlibs** [comment]
@psrinivasa2 , yes, I was referring to assigning logical DB ownership to a particular login - [doc](https://github.com/wiltondb/wiltondb/wiki/Logins-and-users#creating-login-owned-databases).

**Comment by psrinivasa2** [comment]
this answer's, thanks Alex @staticlibs

**Comment by staticlibs** [comment]
@psrinivasa2 , I would think the problem is not related to authentication. Just something is wrong with some of the SQL queries that are issued by SSMS after the connection is established. Can you copy the relevant bits from DB server log when the problem happens? Also, what is the exact version of SSMS you use?

**Comment by psrinivasa2** [comment]
Hi @staticlibs,
Please find the attached screenshot

![image](https://github.com/user-attachments/assets/33b6dafb-b767-4833-8335-10b177f2c914)

![image](https://github.com/user-attachments/assets/09ca9a26-a6da-4dd5-98f0-9486ba76b763)

**Comment by staticlibs** [comment]
@psrinivasa2 , I filed #35 to track this "Invalid cast" problem with SSMS, please leave your comments there, if necessary.

**Comment by psrinivasa2** [comment]
Hi Alex @staticlibs ,
After I create the OS user for myself, I don't see the use for 'wilton' user. We are maintaining 5000 clients, difficult to maintain the custom wilton user password for all of them, is it possible to remove 'wilton' user by running any query after creating OS user?

![image](https://github.com/user-attachments/assets/3ca1276f-6bf3-4317-9275-95093f8d6dbc)

**Comment by staticlibs** [comment]
@psrinivasa2 , perhaps the easiest way to disable `wilton` user would be to change the password for this user (on Postgres connection) to some random unknown one (for example to `gen_random_uuid()` output):

```sql
alter user wilton with password '...'
```
